regen/embed.pl: Fix, refactor ARGS_ASSERT generation

khwilliamson opened the pull request:
This fixes a bug in which certain conditions failed to generate a required ARGS_ASSERT macro

It refactors the generation in preparation for future commits and changes the embed.fnc G flag effect so that it is overridden for all cases where a non-empty ARGS_ASSERT entry is generated; previously only some cases overrode it.

* This set of changes does not require a perldelta entry.

khwilliamson committed:
embed.fnc: Synchronize formal parameter name with actual

A PERL_ARGS_ASSERT macro will be generated for this function in a future
commit, and compilation will fail unless the names match.

khwilliamson committed:
regen/embed.pl: Generate ARGS_ASSERT in more cases

This macro is actually always generated, but is empty if there are no
assertions needed.

Before this commit the code incorrectly generated an empty macro when
there were no arguments marked NN even when types were defined that
would call for asserts to be generated.

khwilliamson committed:
Add some ARGS_ASSERT calls

An ARGS_ASSERT macro is always generated for every function listed in
embed.fnc, unless possibly suppressed with the G flag for that entry.
The generated macro is empty if there is nothing to assert.

It is mandatory (enforced through a porting test) to call that macro
when non-empty.  (Hopefully the call occurs in the function the macro is
designed for, but the porting test is currently simplistic and doesn't
check for that; often compilation would fail anyway if it did get placed
in the wrong function, as the parameter names the macro expects and the
ones in the function could easily not match).

It is optional (but a good idea) to call the macro even when empty.
That way this commit would not have been necessary.  From time to time,
an empty macro becomes non-empty as we figure out more things to check
for.  When that happens, the porting test fails for the newly-non-empty
macros that aren't called.  If the function had originally called the
empty-one, its source wouldn't have to change at all.

Several commits from now will make some ARGS_ASSERT macros non-empty;
this commit adds calls to the ones that weren't already called.

khwilliamson committed:
embed.fnc: Add NOCHECK to three entries

This has no effect currently, but several commits hence, a bug will be
fixed that would otherwise generate a check that actually is
inappropriate for these cases.

bram-perl commented on regen/embed.pl:
```diff
@@ -355,7 +357,7 @@ sub generate_proto_h {
         $ret = "/* $ret */" if $has_mflag;
 
-        if ($args_assert_line || @names_of_nn) {
+        if ($args_assert_line) {
```
**Commit regen/embed.pl: Honor -G flag**

A bit of nit-picking in the commit message:
> This flag says don't generate a PERL_ARGS_ASSERT line. 

I guess it would be more correct to say:
> This flag says don't generate a PERL_ARGS_ASSERT **macro**. 

Also looking at [embed.fnc](https://github.com/Perl/perl5/blob/e31d3ed5601e07f26b798df942c1d874c68897f3/embed.fnc#L391)
```
:   'G'  Suppress empty PERL_ARGS_ASSERT_foo macro. Normally such a macro is
:        generated for all entries for functions 'foo' in this file. If there
:        is a pointer argument to 'foo', it needs to be declared in this file
:        as either NN or NULLOK, and the function definition must call its
:        corresponding PERL_ARGS_ASSERT_foo macro (a porting test ensures this)
:        which asserts at runtime (under DEBUGGING builds) that NN arguments
:        are not NULL. If there aren't NN arguments, use of this macro is
:        optional. Rarely, a function will define its own PERL_ARGS_ASSERT_foo
:        macro, and in those cases, adding this flag to its entry in this file
:        will suppress the normal one. It is not possible to suppress the
:        generated macro if it isn't optional, that is, if there is at least
:        one NN argument.
:
:          proto.h: PERL_ARGS_ASSERT macro is not defined unless the function
:                   has NN arguments
```

but reading that it doesn't really match with how "-G" is described in the commit message. (And I guess the embed.fnc doc needs updating):
- the docs: supress **empty** PERL_ARGS_ASSERT
- the commit: supress PERL_ARGS_ASSERT in all cases

khwilliamson replied:
Thank you for your detailed review

I changed the G flag handling because it looked wrong to me.  Now that I know it was intentional and documented as such, I believe the only change is that any assertion generated will override the G flag

bram-perl replied:
Should the "-G" docs be updated?

I guess the old doc was also incomplete: the -G would be ignored if there were NN **or** NZ arguments.
And in the new code it's when there are NN **or** NZ **or** type asserts.

(Alternative: drop the -G flag entirely? I had a quick look earlier and it's only used for ~2~4 functions? Not sure if that is really worth the effort of having a separate flag flag for it and/or if it's even useful for those functions..) <details>
    <summary>Functions with -G</summary>

```
CGTdp   |void   |noshutdownhook

EGdp    |HV *   |gv_stashsvpvn_cached                                   \
                                |SV *namesv                             \
                                |const char *name                       \
                                |U32 namelen                            \
                                |I32 flags


GS      |bool   |isa_lookup     |NULLOK HV *stash                       \
                                |NULLOK SV *namesv                      \
                                |NULLOK const char *name                \
                                |STRLEN len                             \
                                |U32 flags

GS      |bool   |sv_derived_from_svpvn                                  \
                                |NULLOK SV *sv                          \
                                |NULLOK SV *namesv                      \
                                |NULLOK const char *name                \
                                |const STRLEN len                       \
                                |U32 flags
```
</details>

khwilliamson replied:
G flag update is in #23513

bram-perl commented on regen/embed.pl:
```diff
@@ -359,7 +359,7 @@ sub generate_proto_h {
         if ($args_assert_line) {
             $ret .= "\n#${ind}define PERL_ARGS_ASSERT_\U$plain_func\E";
-            if (@names_of_nn) {
+            if (@names_of_nn || @typed_args) {
```
**Commit regen/embed.pl: Generate ARGS_ASSERT in more cases**

In the message:
> But unntil this commit the code could wrongly create an empty macro.
> This happened unless at least one of the arguments to the function was
> marked NN.

It contains a small typo (unntil  -> until) but I also find the message a bit difficult to read.
(I would also replace *could* with *would*)

Generally I think I would write something like:
```
Before this commit the code incorrectly generated an empty macro when there were no arguments marked NN even when types were defined.
```
(or something like)

khwilliamson replied:
done

bram-perl commented on regen/embed.pl:
```diff
@@ -274,13 +273,25 @@ sub generate_proto_h {
+
+                if (   $args_assert_line
+                    && defined $argname
+                    && ! ($has_mflag && ! $binarycompat))
```
**Commit regen/embed.pl: Refactor ARGS_ASSERT generation**

I know this line was copied from the old code but I don't like it very much.
Personally I would find it cleaner if it was written as: `&& (! $has_mflag || $binarycompat))` to avoid a double negative

bram-perl commented on regen/embed.pl:
```diff
@@ -274,13 +273,25 @@ sub generate_proto_h {
+
+                if (   $args_assert_line
+                    && defined $argname
+                    && ! ($has_mflag && ! $binarycompat))
```
**Commit regen/embed.pl: Refactor ARGS_ASSERT generation**

With the old code: when adding items to `@typed_args` it did not check for `defined $argname` nor `! ($has_mflag && ! $binarycompat)`. An assert would be added anyway.

With the new code: asserts are only added when `defined $argname` and `! ($has_mflag && ! $binarycompat)` are  true.

What I'm trying to get at: with the old code there could (or maybe are?) asserts when:
- `$argname` is not defined
- `! ($has_mflag && ! $binarycompat)` is false

with the new code that will never be the case.

Ideally I would like see some explanation for that in the commit message (on why that change is fine: is it something that can't/shouldn't happen?).

khwilliamson replied:
Split into tiny commits with explanations

khwilliamson committed:
regen/embed.pl: Rm double negative in conditional

As suggested by Bram, this was hard to understand

khwilliamson committed:
regen/embed.pl: Add a condition to an 'if'

$argname is used inside this 'if' as part of a newly created array
element.  It needs to be defined for the code to work that later looks
at that element.

No current calls result in an undefined value for this.

khwilliamson committed:
regen/embed.pl: Add a condition to an 'if'

This is in preparation of combining this if clause with the one just
below to as many of the conditions in common as possible.

In this clause, it applies only to functions, as we don't create
ARGS_ASSERT macros for macros.  However when something is going into
mathoms.c (b flag), it is marked as a macro, but also has a function, so
we need to create the ASSERT in this case.

khwilliamson committed:
regen/embed.pl: Swap the order of two blocks

This makes no difference in processing now, but will minimize the 'diff'
output of a future commit.

khwilliamson committed:
regen/embed.pl: Combine the common parts of two 'if's

This commit combines two separate 'if' blocks into one big if block with
the conditionals in common to the two blocks, and then the two smaller
blocks, each with the conditions that apply to it individually.

And then everything is reindented to correspond

bram-perl commented on regen/embed.pl:
```diff
@@ -277,14 +276,18 @@ sub generate_proto_h {
-                        push @typed_args, [ $argtype, $argname ];
+                        my $type_assert =
+                            $type_asserts{$argtype} =~ s/__arg__/$argname/gr;
+                        push @asserts,
+                            $nullok ? "assert(!$argname || $type_assert)"
```
**Commit regen/embed.pl: Refactor ARGS_ASSERT generation**

This is a tricky one..
It was moved from the old block and that also had a `$nullok` variable and so does this block but technically they are not the same (maybe practically they are?)

Relevant pieces of the old code:
```
my $nn = ( $arg =~ s/\s*\bNN\b\s+// );
my $nz = ( $arg =~ s/\s*\bNZ\b\s+// );
...
if ($nn||$nz) {
    push @names_of_nn, $argname;
}
...
my $nullok = !grep { $_ eq $argname } @names_of_nn;
```

Relevant pieces of the new code:
```
my $nullok = ( $arg =~ s/\s*\bNULLOK\b\s+// ); # strip NULLOK with no effect
```

**TLDR**:
- old code: `$nullok` is basically arg not containing `NN` **and** not containing `NZ`
- new code: `$nullok` is the arg containing `NULLOK`


For NULLOK; the comment behind it ` # strip NULLOK with no effect`: I wasn't entirely sure what it really meant so looking at [embed.fnc](https://github.com/Perl/perl5/blob/463db482666ccf830e8eb7ab26052dccf4991176/embed.fnc#L629)
```
: And, pointer parameters that may be NULL should be prefixed with NULLOK.
: This has no effect on output yet.  It's a notation for the maintainers to
: know "I have defined whether NULL is OK or not" rather than having neither
: NULL or NULLOK, which is ambiguous.
```
This is no longer true, it now does have an effect on the output.

(I have no relevant opinion on the change only that if the change is made the doc should be updated.)

khwilliamson replied:
#23513 changes the text describing NULLOK, so that problem doesn't come up

It is somewhat subtler still.  embed.fnc does not have extensive error checking; it is assumed you aren't going to screw up.  Legal input has NN and  NULLOK only on pointer arguments; and NZ only on non pointers.  That changes the tl;dr above to

-    old code:  $nullok is basically a pointer arg not containing NN or a non-pointer not containing NZ
-    new code: $nullok is a pointer arg containing NULLOK

Error checking is currently done enough to make sure all pointer args are either NN or NULLOK.  That means a pointer arg not containing NN must contain NULLOK.  This allows these to be rewritten as

-    old code:  $nullok is a pointer arg  containing NULLOK or a non-pointer not containing NZ
-    new code: $nullok is a pointer arg containing NULLOK

For both pieces of code, a pointer arg containing NULLOK outputs nothing.

For the old code, a non-pointer not containing NZ also outputs nothing.  That's also true of the new code, so the missing clause is irrelevant.

khwilliamson replied:
I have now split this commit, with an explanation in one of the splits' message

khwilliamson committed:
regen/embed.pl: Get rid of ternary

This simple change makes the code a bit easier to understand.

khwilliamson committed:
regen/embed.pl Move some code to earlier

Prior to this commit, the code created an array describing an input
argument, adding it to a list of all the arguments, and later went
through that list using a grep to reconstruct some information that had
been lost (because the array didn't include all the needed relevant
information), and placed the result in its final form.

This commit moves the calculation of the final form to earlier, where it
first is encountered.  This means all the information is available, and
no grep is needed.

Code review revealed a subtlety here.  The moved code has an

    if $nullok

but prior to this commit, there were two different $nullok variables,
with slightly different meanings.  Prior to this commit, the moved code
was using the second $nullok;  afterwards, the first one.  Hence, even
though the name is the same, before and after, it is a different
variable, with a (slightly) different meaning.

The second $nullok was populated by the grep, and meant that the
argument did not have either a NN nor a NZ modifier.

The first $nullok meant simply that the arg had a NULLOK modifier.

After the commit the single $nullok means that the arg had a NULLOK
modifier.

This change actually has no effect (as demonstrated by the fact that
this commit doesn't change the generated proto.h).  The moved code is
executed only for an argument where $type_asserts{$argtype} exists.
This hash is constant, and has been populated so that only pointer
arguments have entries.  NZ is not valid for pointer arguments, so we
know that both before and after this commit, the arg did not have an NZ
modifier.  Also it is illegal to specify both NN and NULLOK, so both
before and after this commit, the arg did not have a NN modifier.  And
both before and after this commit the arg does have a NULLOK modifier.

khwilliamson committed:
regen/embed.pl: Combine two arrays

These two arrays are combined anyway into a third array later in the
code; this just goes with the third array from the start.

The order of some assertions in the generated proto.h are changed.  But
this makes sure that we continue to assert that an argument is present
before dereferencing it.

khwilliamson committed:
regen/embed.pl: Remove a loop

By creating the @asserts array with each element in its final form, the
loop that later did this can be eliminated.

khwilliamson committed:
regen/embed.pl: White-space/comment changes

This removes a misleading comment.  NULLOK now does have some effect.
It changes the actual generated assertion for arguments where a type
assertion is generated.  (This has actually been true since
2463f19365f941f68e9d5eed2f787341df8ccdef)

This also vertically aligns some statements for ease of reading, and
reorders things to group them together in a block.
